File: packages/lib/src/__tests__/quiz/pyq-years.test.ts
/**
 * PYQ board-year window + validator.
 *
 * The retired `/pyq` runtime hardcoded `2025 - i` for 11 entries. That list
 * decays silently: past its author's assumption it simply stops offering the
 * newest board paper and nothing fails. These tests pin the derived window and,
 * more importantly, pin that the PICKER and the `?year=` VALIDATOR agree — if
 * they drift, `/pyq` offers a year that `/quiz` then discards, and the student
 * gets a generic quiz labelled as a board paper.
 */

import { describe, it, expect } from 'vitest';
import {
  PYQ_MIN_YEAR,
  PYQ_YEAR_COUNT,
  isPyqYear,
  pyqMaxYear,
  pyqYears,
} from '../../quiz/pyq-years';

const AT = (iso: string) => new Date(iso);

describe('pyqYears — the offered window', () => {
  it('is newest-first and PYQ_YEAR_COUNT long', () => {
    const years = pyqYears(AT('2026-08-11T00:00:00Z'));
    expect(years).toHaveLength(PYQ_YEAR_COUNT);
    expect(years[0]).toBe(2026);
    expect(years.at(-1)).toBe(2016);
    expect([...years].sort((a, b) => b - a)).toEqual(years);
  });

  it('advances with the calendar instead of decaying (the defect it replaces)', () => {
    expect(pyqYears(AT('2027-01-01T00:00:00Z'))[0]).toBe(2027);
    expect(pyqYears(AT('2030-06-01T00:00:00Z'))[0]).toBe(2030);
  });

  it('never offers a paper that has not been sat yet', () => {
    // Board papers for year Y are sat in Feb-Mar of Y, so Y itself is the
    // newest honest offer; Y+1 would be a promise content cannot keep.
    const now = AT('2026-08-11T00:00:00Z');
    expect(Math.max(...pyqYears(now))).toBe(pyqMaxYear(now));
    expect(pyqYears(now)).not.toContain(2027);
  });

  it('never falls below the floor', () => {
    const years = pyqYears(AT('2015-01-01T00:00:00Z'));
    expect(Math.min(...years)).toBeGreaterThanOrEqual(PYQ_MIN_YEAR);
  });
});

describe('isPyqYear — the ?year= gate', () => {
  const now = AT('2026-08-11T00:00:00Z');

  it('accepts every year the picker offers (picker and gate cannot drift)', () => {
    for (const y of pyqYears(now)) expect(isPyqYear(y, now)).toBe(true);
  });

  it('accepts an older bookmarked year down to the floor', () => {
    // Deliberately WIDER than the picker window: an old link keeps working
    // instead of quietly degrading to a generic quiz.
    expect(isPyqYear(PYQ_MIN_YEAR, now)).toBe(true);
    expect(pyqYears(now)).not.toContain(PYQ_MIN_YEAR);
  });

  it.each([
    ['below the floor', PYQ_MIN_YEAR - 1],
    ['in the future', 2027],
    ['junk small integer', 3],
    ['a year-like float', 2019.5],
    ['NaN', Number.NaN],
    ['Infinity', Number.POSITIVE_INFINITY],
  ])('rejects %s', (_label, value) => {
    expect(isPyqYear(value, now)).toBe(false);
  });

  it.each([
    ['a numeric string', '2019'],
    ['null', null],
    ['undefined', undefined],
    ['an object', {}],
  ])('rejects %s (non-number)', (_label, value) => {
    expect(isPyqYear(value, now)).toBe(false);
  });
});

File: mobile/test/core/pyq_years_test.dart
// Tests for the clock-derived PYQ board-year window (R8).
//
// The point of this module is that it does NOT decay, so the tests pin the
// behaviour that a hardcoded list cannot have: the window MOVES with the clock.
// Constants are asserted against `packages/lib/src/quiz/pyq-years.ts` so the
// two runtimes cannot silently drift apart and offer different papers.

import 'package:alfanumrik/core/constants/pyq_years.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('constants match the web module', () {
    test('count and floor mirror pyq-years.ts', () {
      // PYQ_YEAR_COUNT = 11, PYQ_MIN_YEAR = 2010 in packages/lib.
      expect(kPyqYearCount, 11);
      expect(kPyqMinYear, 2010);
    });
  });

  group('pyqMaxYear', () {
    test('is the current year, never the next one', () {
      // Boards for year Y are sat in Feb-Mar of Y, so Y is offerable within Y.
      // `+1` would promise a paper that has not been written.
      expect(pyqMaxYear(now: DateTime.utc(2026, 8, 11)), 2026);
      expect(pyqMaxYear(now: DateTime.utc(2026, 12, 31, 23, 59)), 2026);
    });

    test('reads UTC, matching the web getUTCFullYear()', () {
      // 1 Jan 2027 00:30 IST is still 31 Dec 2026 UTC. Web and mobile must
      // agree on which side of the boundary this is.
      final istNewYear = DateTime.utc(2026, 12, 31, 19, 0); // 00:30 IST, 1 Jan
      expect(pyqMaxYear(now: istNewYear), 2026);
    });
  });

  group('pyqYears', () {
    test('returns kPyqYearCount years, newest first', () {
      final years = pyqYears(now: DateTime.utc(2026, 6, 1));
      expect(years.length, kPyqYearCount);
      expect(years.first, 2026);
      expect(years.last, 2016);
      // Strictly descending, no gaps.
      for (var i = 1; i < years.length; i++) {
        expect(years[i], years[i - 1] - 1);
      }
    });

    test('WINDOW MOVES with the clock — the decay the literal had', () {
      // This is the whole reason the hardcoded `[2025 ... 2015]` list was
      // deleted: once the calendar passed its author's assumption it silently
      // stopped offering the newest board paper and nothing failed.
      final y2026 = pyqYears(now: DateTime.utc(2026, 1, 1));
      final y2031 = pyqYears(now: DateTime.utc(2031, 1, 1));
      expect(y2026.first, 2026);
      expect(y2031.first, 2031);
      expect(y2031.contains(2031), isTrue);
      expect(y2026.contains(2031), isFalse);
    });

    test('never offers a year below the floor', () {
      final years = pyqYears(now: DateTime.utc(2015, 1, 1));
      expect(years.every((y) => y >= kPyqMinYear), isTrue);
      expect(years.last, kPyqMinYear);
      expect(years.length, lessThan(kPyqYearCount));
    });

    test('is unmodifiable so a caller cannot mutate the shared window', () {
      expect(() => pyqYears().add(1999), throwsUnsupportedError);
    });
  });

  group('isPyqYear', () {
    final now = DateTime.utc(2026, 6, 1);

    test('accepts the full floor..max range, not just the visible window', () {
      // An older bookmarked deep link keeps working rather than quietly
      // degrading to a generic quiz.
      expect(isPyqYear(2026, now: now), isTrue);
      expect(isPyqYear(2016, now: now), isTrue);
      expect(isPyqYear(2010, now: now), isTrue); // outside pyqYears(), still ok
    });

    test('rejects junk and future papers', () {
      expect(isPyqYear(null, now: now), isFalse);
      expect(isPyqYear(3, now: now), isFalse);
      expect(isPyqYear(1970, now: now), isFalse);
      expect(isPyqYear(2009, now: now), isFalse);
      expect(isPyqYear(2027, now: now), isFalse);
    });
  });

  group('parsePyqYear (the deep-link boundary)', () {
    final now = DateTime.utc(2026, 6, 1);

    test('parses a valid ?year=', () {
      expect(parsePyqYear('2019', now: now), 2019);
      expect(parsePyqYear(' 2019 ', now: now), 2019);
    });

    test('returns null for hostile or absent input rather than throwing', () {
      // A hostile deep link must never reach a question filter, and must not
      // crash the route builder either — it degrades to a normal quiz.
      expect(parsePyqYear(null, now: now), isNull);
      expect(parsePyqYear('', now: now), isNull);
      expect(parsePyqYear('abc', now: now), isNull);
      expect(parsePyqYear('2019; DROP TABLE', now: now), isNull);
      expect(parsePyqYear('1970', now: now), isNull);
      expect(parsePyqYear('2099', now: now), isNull);
      expect(parsePyqYear('-2019', now: now), isNull);
    });
  });
}
